include/logging/log_msg2.h

```diff
@@ -34,6 +34,25 @@ typedef uint64_t log_timestamp_t;
 typedef uint32_t log_timestamp_t;
 #endif
 
+/* Determine if auto/__auto_type is supported. If not then runtime approach must always
+ * be used.
+ *
+ * Supported by:
+ * - C++ (auto)
+ * - GCC 4.9.0 https://gcc.gnu.org/gcc-4.9/changes.html
+ * - Clang 3.8
+ */
+#if defined(__cplusplus)
+#define Z_AUTO_TYPE auto
+#elif ((__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__) >= 40900) || \
+	((__clang_major__ * 10000 + __clang_minor__ * 100 + __clang_patchlevel__) >= 30800)
+#define Z_AUTO_TYPE __auto_type
+#else
+#ifndef CONFIG_LOG2_ALWAYS_RUNTIME
+#error "Compiler lacks support for __auto_type. Enable CONFIG_LOG2_ALWAYS_RUNTIME."
```

Comment:
Never got a reply on this part:
https://github.com/zephyrproject-rtos/zephyr/pull/42686#discussion_r808932898
> For example checking compiler version in CMake

This can be done with: https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_VERSION.html which is supported for gcc and llvm.
That way we may detect this BEFORE compilation.

Reply:
that way we can inform user earlier that there is a configuration issue that must be fixed.